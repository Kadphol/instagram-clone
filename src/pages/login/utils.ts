export const validateEmail = (email: string) => {
  const regex = /\S+@\S+\.\S+/
  if (email.length === 0) {
    return 'Email is required'
  } else if (!regex.test(email)) {
    return 'Please use a valid email address!'
  } else if (email.length > 30) {
    return 'Email address cannot exceed more than 30 characters'
  }

  return ''
}

export const validatePassword = (password: string) => {
  const regexLetters = /^(?=.*[a-z])(?=.*[A-Z])/
  const regexNumbers = /^(?=.*[0-9])/

  if (password.length === 0) {
    return 'Password is required'
  } else if (!regexLetters.test(password)) {
    return 'Password must contain at least one lowercase and uppercase letter'
  } else if (!regexNumbers.test(password)) {
    return 'Password must contain at least one number'
  } else if (password.length < 8) {
    return 'Password be must be eight characters or longer'
  }

  return ''
}
