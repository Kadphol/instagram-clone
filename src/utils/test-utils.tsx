import { render } from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime'
import { NextRouter } from 'next/router'

type DefaultParams = Parameters<typeof render>
type RenderUI = DefaultParams[0]
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> }

const mockRouter: NextRouter = {
  basePath: '',
  pathname: '/',
  route: '/',
  asPath: '/',
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  forward: jest.fn(),
  isReady: false,
  isPreview: false,
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
  isLocaleDomain: false,
}

export const createMockRouter = (router: Partial<NextRouter>): NextRouter => {
  return {
    basePath: '',
    pathname: '/',
    route: '/',
    query: {},
    asPath: '/',
    back: jest.fn(),
    beforePopState: jest.fn(),
    prefetch: jest.fn(),
    push: jest.fn(),
    reload: jest.fn(),
    replace: jest.fn(),
    forward: jest.fn(),
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    isFallback: false,
    isLocaleDomain: false,
    isReady: true,
    defaultLocale: 'en',
    domainLocales: [],
    isPreview: false,
    ...router,
  }
}

/**  ref: https://wnassour.com/posts/test-router-in-nextjs
 * --------------------------------------------------
 *  Override the default test render with our own
 *
 *  You can override the router mock like this:
 *
 *  const { baseElement } = render(<MyComponent />, {
 *    router: { pathname: '/my-custom-pathname' },
 *  });
 * --------------------------------------------------
 */

export const renderWithRouter = (
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
) => {
  if (!wrapper) {
    wrapper = ({ children }) => (
      <RouterContext.Provider value={{ ...mockRouter, ...router }}>
        {children}
      </RouterContext.Provider>
    )
  }

  return render(ui, { wrapper, ...options })
}
