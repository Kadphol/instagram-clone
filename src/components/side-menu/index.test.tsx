import { createMockRouter, renderWithRouter } from '@/utils/test-utils'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import SideMenu from './index'

describe('<SideMenu />', () => {
  beforeEach(() => {
    jest.restoreAllMocks()
    localStorage.clear()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders the menu items', () => {
    const { asFragment } = renderWithRouter(<SideMenu />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('should open menu when click hamburger icon', async () => {
    const user = userEvent.setup()

    renderWithRouter(<SideMenu />)

    const menuIcon = screen.getByTestId('menu')

    await user.click(menuIcon)

    expect(screen.getByRole('button', { name: 'Log out' })).toBeInTheDocument()
  })

  it('should logout when click logout button', async () => {
    const user = userEvent.setup()
    const router = createMockRouter({ push: jest.fn() })
    const localStorageRemoveItemSpy = jest.spyOn(localStorage, 'removeItem')

    renderWithRouter(<SideMenu />, { router })

    const menuIcon = screen.getByTestId('menu')

    await user.click(menuIcon)

    const logoutButton = screen.getByRole('button', { name: 'Log out' })

    await user.click(logoutButton)

    expect(localStorageRemoveItemSpy).toBeCalledWith('loggedIn')
    expect(localStorageRemoveItemSpy).toBeCalledWith('guest')
    expect(router.push).toBeCalledWith('/login')
  })
})
