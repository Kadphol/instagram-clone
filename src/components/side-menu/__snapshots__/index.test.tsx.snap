// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideMenu /> renders the menu items 1`] = `
<DocumentFragment>
  <aside
    class="fixed hidden h-full w-16 flex-col items-center overflow-y-auto border-r border-gray-light bg-white px-3 py-5 sm:flex"
  >
    <div
      class="mt-3 pb-6"
    >
      <a
        href="/"
      >
        <svg
          fill="currentColor"
          height="1.5rem"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.1 141c-63.6 0-114.9 51.3-114.9 114.9s51.3 114.9 114.9 114.9S339 319.5 339 255.9 287.7 141 224.1 141zm0 189.6c-41.1 0-74.7-33.5-74.7-74.7s33.5-74.7 74.7-74.7 74.7 33.5 74.7 74.7-33.6 74.7-74.7 74.7zm146.4-194.3c0 14.9-12 26.8-26.8 26.8-14.9 0-26.8-12-26.8-26.8s12-26.8 26.8-26.8 26.8 12 26.8 26.8zm76.1 27.2c-1.7-35.9-9.9-67.7-36.2-93.9-26.2-26.2-58-34.4-93.9-36.2-37-2.1-147.9-2.1-184.9 0-35.8 1.7-67.6 9.9-93.9 36.1s-34.4 58-36.2 93.9c-2.1 37-2.1 147.9 0 184.9 1.7 35.9 9.9 67.7 36.2 93.9s58 34.4 93.9 36.2c37 2.1 147.9 2.1 184.9 0 35.9-1.7 67.7-9.9 93.9-36.2 26.2-26.2 34.4-58 36.2-93.9 2.1-37 2.1-147.8 0-184.8zM398.8 388c-7.8 19.6-22.9 34.7-42.6 42.6-29.5 11.7-99.5 9-132.1 9s-102.7 2.6-132.1-9c-19.6-7.8-34.7-22.9-42.6-42.6-11.7-29.5-9-99.5-9-132.1s-2.6-102.7 9-132.1c7.8-19.6 22.9-34.7 42.6-42.6 29.5-11.7 99.5-9 132.1-9s102.7-2.6 132.1 9c19.6 7.8 34.7 22.9 42.6 42.6 11.7 29.5 9 99.5 9 132.1s2.7 102.7-9 132.1z"
          />
        </svg>
      </a>
    </div>
    <div
      class="my-2 flex flex-col py-3"
    >
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <svg
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12.97 2.59a1.5 1.5 0 0 0-1.94 0l-7.5 6.363A1.5 1.5 0 0 0 3 10.097V19.5A1.5 1.5 0 0 0 4.5 21h4.75a.75.75 0 0 0 .75-.75V14h4v6.25c0 .414.336.75.75.75h4.75a1.5 1.5 0 0 0 1.5-1.5v-9.403a1.5 1.5 0 0 0-.53-1.144l-7.5-6.363Z"
            />
          </svg>
        </a>
      </div>
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <svg
            fill="none"
            height="1.5rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </a>
      </div>
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <svg
            aria-label="Reels"
            color="rgb(0, 0, 0)"
            fill="rgb(0, 0, 0)"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <line
              fill="none"
              stroke="currentColor"
              stroke-linejoin="round"
              stroke-width="2"
              x1="2.049"
              x2="21.95"
              y1="7.002"
              y2="7.002"
            />
            <line
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              x1="13.504"
              x2="16.362"
              y1="2.001"
              y2="7.002"
            />
            <line
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              x1="7.207"
              x2="10.002"
              y1="2.11"
              y2="7.002"
            />
            <path
              d="M2 12.001v3.449c0 2.849.698 4.006 1.606 4.945.94.908 2.098 1.607 4.946 1.607h6.896c2.848 0 4.006-.699 4.946-1.607.908-.939 1.606-2.096 1.606-4.945V8.552c0-2.848-.698-4.006-1.606-4.945C19.454 2.699 18.296 2 15.448 2H8.552c-2.848 0-4.006.699-4.946 1.607C2.698 4.546 2 5.704 2 8.552Z"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M9.763 17.664a.908.908 0 0 1-.454-.787V11.63a.909.909 0 0 1 1.364-.788l4.545 2.624a.909.909 0 0 1 0 1.575l-4.545 2.624a.91.91 0 0 1-.91 0Z"
              fill-rule="evenodd"
            />
          </svg>
        </a>
      </div>
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <svg
            fill="currentColor"
            height="1.5rem"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.76497 19.225C8.35411 18.9652 9.01479 18.9164 9.63316 19.0861C10.4032 19.2963 11.198 19.4019 12.001 19.4C16.5861 19.4 20.001 16.1135 20.001 11.7C20.001 7.28651 16.5861 4 12.001 4C7.41585 4 4.00098 7.28651 4.00098 11.7C4.00098 13.9699 4.89652 15.9722 6.46655 17.3764C7.0418 17.8853 7.38251 18.6082 7.409 19.3822L7.76497 19.225ZM12.001 2C17.6345 2 22.001 6.1265 22.001 11.7C22.001 17.2735 17.6345 21.4 12.001 21.4C11.0233 21.4023 10.0497 21.273 9.10648 21.0155C8.92907 20.9668 8.7403 20.9808 8.57198 21.055L6.58748 21.931C6.34398 22.0386 6.06291 22.018 5.83768 21.8761C5.61244 21.7342 5.47254 21.4896 5.46448 21.2235L5.40998 19.4445C5.40257 19.2257 5.30547 19.0196 5.14148 18.8745C3.19598 17.1345 2.00098 14.6155 2.00098 11.7C2.00098 6.1265 6.36748 2 12.001 2ZM5.99598 14.5365L8.93348 9.8765C9.15689 9.5221 9.51834 9.27728 9.93034 9.2013C10.3423 9.12532 10.7673 9.22511 11.1025 9.4765L13.439 11.2265C13.6528 11.3878 13.9476 11.3878 14.1615 11.2265L17.317 8.8315C17.738 8.512 18.288 9.016 18.006 9.4635L15.0685 14.1235C14.8451 14.4779 14.4836 14.7227 14.0716 14.7987C13.6596 14.8747 13.2346 14.7749 12.8995 14.5235L10.563 12.7735C10.3491 12.6122 10.0543 12.6122 9.84048 12.7735L6.68498 15.1685C6.26398 15.488 5.71398 14.984 5.99598 14.5365Z"
            />
          </svg>
        </a>
      </div>
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <svg
            fill="none"
            height="1.5rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
            />
          </svg>
        </a>
      </div>
      <div
        class="my-2 py-3"
      >
        <a
          href="/"
        >
          <img
            alt="profile image"
            class="rounded-full"
            data-nimg="1"
            decoding="async"
            height="24"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F24%2F24&w=48&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fpicsum.photos%2F24%2F24&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fpicsum.photos%2F24%2F24&w=48&q=75 2x"
            style="color: transparent;"
            width="24"
          />
        </a>
      </div>
    </div>
    <div
      class="mt-auto flex"
    >
      <div
        class="my-2 py-3"
        data-testid="menu"
      >
        <svg
          fill="none"
          height="1.5rem"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 15 15"
          width="1.5rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </aside>
</DocumentFragment>
`;
